lib/install/proto/client.go

```diff
@@ -28,35 +28,77 @@ import (
 )
 
 // NewClient returns a new client using the specified socket file path
-func NewClient(ctx context.Context, socketPath string, logger log.FieldLogger, opts ...grpc.DialOption) (AgentClient, error) {
+func NewClient(ctx context.Context, config ClientConfig) (AgentClient, error) {
 	dialOptions := []grpc.DialOption{
 		// Don't use TLS, as we communicate over domain sockets
 		grpc.WithInsecure(),
 		// Retry every second after failure
 		grpc.WithBackoffMaxDelay(1 * time.Second),
 		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
 		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
-			conn, err := (&net.Dialer{}).DialContext(ctx, "unix", socketPath)
-			logger.WithFields(log.Fields{
+			if err := config.IsServiceFailed(); err != nil {
+				// Fast path: service has failed
+				// Note, the error is not trace.Wrapped on purpose - it needs
+				// to retain the fact that it's terminal
+				return nil, serviceError(trace.UserMessage(err))
+			}
+			conn, err := (&net.Dialer{}).DialContext(ctx, "unix", config.SocketPath)
+			config.WithFields(log.Fields{
 				log.ErrorKey: err,
-				"addr":       socketPath,
+				"addr":       config.SocketPath,
 			}).Debug("Connect to installer service.")
 			if err != nil {
 				return nil, trace.Wrap(err)
 			}
 			return conn, nil
 		}),
 	}
-	dialOptions = append(dialOptions, opts...)
+	dialOptions = append(dialOptions, config.DialOptions...)
 	conn, err := grpc.DialContext(ctx, "unix:///installer.sock", dialOptions...)
 	if err != nil {
+		if wrappedErr, ok := trace.Unwrap(err).(wrappedError); ok {
```

Comment:
Need this abomination to detect whether the error from Dial is `transport.ConnectionError` (so it can be formatted in a friendlier way) which is in an internal package.